### PR TITLE
Fix schedule:file file extension case match issue

### LINKED ISSUE #9509: Schedule:File failing to detect file types correctly when the file extension contains capitalized character
Issue overview
--------------
When specifying a schedule file, if in the IDF block the file name’s extension contains any capitalized character (not all-lower case extension file names), the simulation will fail with a severe error. For example, in the 5ZoneTDV.idf example, the simulation will fail with a sever error is the file name is given as “TDV_2008_kBtu_CT206.CSV” in the following block:
```
  Schedule:File,
    elecTDVfromCZ06com,      !- Name
    Any Number,              !- Schedule Type Limits Name
    TDV_2008_kBtu_CTZ06.CSV,  !- File Name
    1,                       !- Column Number
    4,                       !- Rows to Skip at Top
    8760,                    !- Number of Hours of Data
    Comma,                   !- Column Separator
    No,                      !- Interpolate to Timestep
    60,                      !- Minutes per Item
    No;                      !- Adjust Schedule for Daylight Savings
```
It will pass if the file name is “TDV_2008_kBtu_CTZ06.csv” (lower case extension) instead. Supposely, the simulation will also fail if the extension name contains any upper case character(s), such as ““TDV_2008_kBtu_CTZ06.Csv”, “TDV_2008_kBtu_CTZ06.cSv”, or “TDV_2008_kBtu_CTZ06.cSV”.

It was discovered that the file type search function to match the file types for this part match on lower lower-case characters—more precisely in the following way only (note that `epJSON` is then a mixed way and the extension string has to strictly match one of the following in a case-sensitive manner:
```
    static constexpr std::array<std::string_view, static_cast<std::size_t>(FileTypes::Num)> FileTypesExt{
        "epJSON", "json", "glhe", "cbor", "msgpack", "ubjson", "bson", "idf", "imf", "csv", "tsv", "txt", "eso", "mtr"};
```

This is occurring for version 22.1 and the most recent develop branch. V9.4 does not have the problem. So the issue should have been introduced somewhere in between v9.4 and v22.1. The current workaround is to change the file extension name to be all-lower case.

Some code changes are needed to fix the issue. This could also possibly affect a few other places that needs determine a file type by extension. One possible way to fix to fix the schedule:file one is included in the draft PR that will follow this issue post shortly. 

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): newer versions such as v22.1; older versions such as v9.4 is not affected.
 - Helpdesk ticket number: discovered while working on an idf file sent along with Ticket 16191 (the ticket itself is about another issue that is not relevant to this post).

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9509))
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #9509 
 - The code fixes the issue that in a Schedule:File input, if the file extension name contains any capital letter(s) (e.g. .CSV, .cSV, or .Csv), the simulation will fail with a severe error message saying the file type is not valid.
 - Current fix tried to convert to all-upper cases to compare. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/FileSystem.cc [hunk 1]
```diff
@@ -223,7 +223,7 @@ namespace FileSystem {
 #endif
 
         extension.remove_prefix(extension.find_last_of('.') + 1);
-        return static_cast<FileTypes>(getEnumerationValue(FileTypesExt, extension));
+        return static_cast<FileTypes>(getEnumerationValueUpper(FileTypesExt, extension));
```
Comment (amirroth) [inline src/EnergyPlus/FileSystem.cc]: That's not the way to do this. The way to do it is to make a `FileTypeExtNamesUC` array where the extensions are upper-case and then to call `getEnumerationValue` with that and `MakeUPPERCase(extension)`. This way, you call `MakeUPPERCase` exactly once. The way you've implemented things here `MakeUPPERCase gets called *twice on every comparison*.

Comment (amirroth) [reply src/EnergyPlus/FileSystem.cc]: One more thing is that you should move `FileTypesExt` and `FileTypesExtUC` to the .cc file since they are only used in this module.

Comment (jcyuan2020) [reply src/EnergyPlus/FileSystem.cc]: Thank! The code has been updated per the comments.

File: src/EnergyPlus/FileSystem.cc [hunk 2]
```diff
@@ -80,6 +80,17 @@ namespace FileSystem {
     std::string const exeExtension;
 #endif
 
+    static constexpr std::array<std::string_view, static_cast<std::size_t>(FileTypes::Num)> FileTypesExt{
+        "epJSON", "json", "glhe", "cbor", "msgpack", "ubjson", "bson", "idf", "imf", "csv", "tsv", "txt", "eso", "mtr"};
+    static constexpr std::array<std::string_view, static_cast<std::size_t>(FileTypes::Num)> FileTypesExtUC{
+        "EPJSON", "JSON", "GLHE", "CBOR", "MSGPACK", "UBJSON", "BSON", "IDF", "IMF", "CSV", "TSV", "TXT", "ESO", "MTR"};
```
Comment (amirroth) [inline src/EnergyPlus/FileSystem.cc:86]: 👍

File: src/EnergyPlus/FileSystem.cc [hunk 3]
```diff
@@ -223,7 +234,7 @@ namespace FileSystem {
 #endif
 
         extension.remove_prefix(extension.find_last_of('.') + 1);
-        return static_cast<FileTypes>(getEnumerationValue(FileTypesExt, extension));
+        return static_cast<FileTypes>(getEnumerationValue(FileTypesExtUC, UtilityRoutines::MakeUPPERCase(extension)));
```
Comment (Myoldmopar) [inline src/EnergyPlus/FileSystem.cc]: Since we are just querying the file type, this seems fully reasonable to me.

File: src/EnergyPlus/FileSystem.cc [hunk 4]
```diff
@@ -80,6 +80,17 @@ namespace FileSystem {
...
+    static constexpr std::array<std::string_view, static_cast<std::size_t>(FileTypes::Num)> FileTypesExt{
+        "epJSON", "json", "glhe", "cbor", "msgpack", "ubjson", "bson", "idf", "imf", "csv", "tsv", "txt", "eso", "mtr"};
+    static constexpr std::array<std::string_view, static_cast<std::size_t>(FileTypes::Num)> FileTypesExtUC{
+        "EPJSON", "JSON", "GLHE", "CBOR", "MSGPACK", "UBJSON", "BSON", "IDF", "IMF", "CSV", "TSV", "TXT", "ESO", "MTR"};
+
+    static_assert(FileTypesExt.size() == static_cast<std::size_t>(FileTypes::Num), "Mismatched FileTypes enum and FileTypesExt array.");
+    static_assert(!FileTypesExt.back().empty(), "Likely missing an enum from FileTypes in FileTypesExt array.");
+
+    static_assert(FileTypesExtUC.size() == static_cast<std::size_t>(FileTypes::Num), "Mismatched FileTypes enum and FileTypesExtUC array.");
+    static_assert(!FileTypesExtUC.back().empty(), "Likely missing an enum from FileTypes in FileTypesExtUC array.");
```
Comment (Myoldmopar) [inline src/EnergyPlus/FileSystem.cc:92]: This is a nice touch; I think there are a handful of other places where we have separate but parallel std::array's that should be asserted like this.

File: tst/EnergyPlus/unit/FileSystem.unit.cc [hunk 5]
```diff
@@ -108,6 +108,52 @@ TEST(FileSystem, getAbsolutePath)
     EXPECT_FALSE(absPathName.find(currentDirWithSep) != std::string::npos); // Make sure "./" doesn't appear in absolute path
 }
 
+TEST(FileSystem, getFileType)
+{
+    fs::path filePath1("./schedulefileA.CSV");
+    auto ext = EnergyPlus::FileSystem::getFileType(filePath1);
+    EXPECT_TRUE(EnergyPlus::FileSystem::FileTypes::CSV == static_cast<EnergyPlus::FileSystem::FileTypes>(ext));
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/FileSystem.unit.cc]: `getFileType` is already returning a `FileTypes` instance, you don't need this `static_cast` at all.  You can also make use of the enum comparison function like this if you want: 
```
EXPECT_TRUE(compare_enums(EnergyPlus::FileSystem::FileTypes::CSV, ext));
```

**Review by Myoldmopar — APPROVED** [review]
This is a clean, simple fix, with a great unit test that covers a bunch of conditions.  The unit test does have a tweak that could be made, but I don't think it's worth holding it up just for that.  Assuming this is all clean and no one has any issues, it can merge.  (If other commits are needed for any reason, please go ahead and clean up the unnecessary casts in the unit test.)

**Comment by Myoldmopar** [comment]
CI is **super** unhappy with this branch, so I'll wait until a fix is made.  While the fix is being made, go ahead and clean up the unnecessary casts in the unit test, please.

File: tst/EnergyPlus/unit/FileSystem.unit.cc (hunk 5 above)
Comment (jcyuan2020) [reply tst/EnergyPlus/unit/FileSystem.unit.cc]: Thanks! The code has been updated by removing the static cast conversions. BTW, I also tried the `compare_enums()` way you suggested (actually first tried this) but it seems to have some problems in recognizing the function `compare_enums()`--although I can see it is defined in `fixtures/EnergyPlusFixuture.hh`. I tried a few combinations by including fixutures/EnergyPlusFixture.hh, and also changing test parameter from `FileSystem` to `EnergyPlusFixture` but none of them would get the function recognized properly here. So finally just stay with the simpler way of removing the unnecessary static cast.

**Comment by Myoldmopar** [comment]
It's still not working.  The Clang build failure seems to [show](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-97c7c29e50237399079b8336f75a3536f460d4ec-PR9510-x86_64-MacOS-10.17-clang-13.0.0.html) it clearly:

```
Undefined symbols for architecture x86_64:
"EnergyPlus::UtilityRoutines::MakeUPPERCase(std::__1::basic_string_view >)", referenced from:
EnergyPlus::FileSystem::getFileType(std::__1::__fs::filesystem::path const&) in libenergyplusparser.a(FileSystem.cc.o)
```

**Comment by Myoldmopar** [comment]
@jcyuan2020 

OK, so there was something _very_ weird going on with the `string_view::remove_prefix` method.  It was like it was properly moving the starting pointer, but the end pointer was being nudged along with it, so the file extension it was checking was like `dfF`.  I think it had something to do with the mix of `string_view` and `fs::path` objects.  As a test, I just did a one-time conversion to a plain old string for the extension, and operated on that.  Everything worked fine after that point.  This code is only being hit very early in the program and just once per file type check, so it's not like this is going to be having any effect on performance.  If CI is happy with this, I think this should be good to go.

**Comment by Myoldmopar** [comment]
Well, I forgot to pull in upstream develop, so this is showing the diffs from https://github.com/NREL/EnergyPlus/pull/9522.  But otherwise it looks clean.  I think this will merge down once integration tests are done and pass fine.

**Comment by Myoldmopar** [comment]
All good here, diffs only due to being out of date with develop.  All files passing.  Merging this, thanks @jcyuan2020
